client/src/pubsub_client_async.rs

```diff
@@ -0,0 +1,323 @@
+use {
+    crate::{
+        rpc_config::{
+            RpcAccountInfoConfig, RpcBlockSubscribeConfig, RpcBlockSubscribeFilter,
+            RpcProgramAccountsConfig, RpcSignatureSubscribeConfig, RpcTransactionLogsConfig,
+            RpcTransactionLogsFilter,
+        },
+        rpc_response::{
+            Response as RpcResponse, RpcBlockUpdate, RpcKeyedAccount, RpcLogsResponse,
+            RpcSignatureResult, RpcVote, SlotInfo, SlotUpdate,
+        },
+    },
+    futures_util::{
+        future::{ready, BoxFuture, FutureExt},
+        sink::SinkExt,
+        stream::{BoxStream, StreamExt},
+    },
+    serde::de::DeserializeOwned,
+    serde_json::{json, Map, Value},
+    solana_account_decoder::UiAccount,
+    solana_sdk::{clock::Slot, pubkey::Pubkey, signature::Signature},
+    std::collections::BTreeMap,
+    thiserror::Error,
+    tokio::{
+        net::TcpStream,
+        sync::{mpsc, oneshot},
+        task::JoinHandle,
+        time::{sleep, Duration},
+    },
+    tokio_stream::wrappers::UnboundedReceiverStream,
+    tokio_tungstenite::{
+        connect_async,
+        tungstenite::{
+            protocol::frame::{coding::CloseCode, CloseFrame},
+            Message,
+        },
+        MaybeTlsStream, WebSocketStream,
+    },
+    url::Url,
+};
+
+pub type PubsubClientResult<T = ()> = Result<T, PubsubClientError>;
+
+#[derive(Debug, Error)]
+pub enum PubsubClientError {
```

Comment:
Can this `PubsubClientError` be shared with the existing enum in `pubsub_client`?